source/common/io/io_uring.h

```diff
@@ -97,24 +116,115 @@ class IoUring {
  */
 class IoUringFactory {
 public:
   virtual ~IoUringFactory() = default;
 
+  virtual OptRef<IoUringWorker> getIoUringWorker() const PURE;
   /**
-   * Returns an instance of IoUring and creates it if needed for the current
-   * thread.
+   * Returns an instance of IoUring for the current thread.
    */
-  virtual IoUring& getOrCreate() const PURE;
+  virtual OptRef<IoUring> get() const PURE;
 
   /**
    * Initializes a factory upon server readiness. For example this method can be
    * used to set TLS.
    */
   virtual void onServerInitialized() PURE;
+
+  virtual bool currentThreadRegistered() PURE;
+};
+
+enum class RequestType { Accept, Connect, Read, Write, Close, Cancel, Unknown };
+
+class IoUringHandler;
+class IoUringSocket;
+
+struct Request {
+  RequestType type_{RequestType::Unknown};
+  absl::optional<std::reference_wrapper<IoUringSocket>> io_uring_socket_{absl::nullopt};
+};
+
+struct AcceptedSocketParam {
+  os_fd_t fd_;
+  sockaddr_storage* remote_addr_;
+  socklen_t remote_addr_len_;
+};
+
+struct ReadParam {
+  Buffer::Instance& pending_read_buf_;
+  int32_t result_;
+};
+
+struct WriteParam {
+  int32_t result_;
+};
+
+class IoUringSocket : public Event::DeferredDeletable {
+public:
+  virtual ~IoUringSocket() = default;
+
+  virtual os_fd_t fd() const PURE;
+  virtual void start() PURE;
+  virtual void close() PURE;
+  virtual void enable() PURE;
+  virtual void disable() PURE;
+
+  virtual uint64_t write(Buffer::Instance&) { PANIC("not implemented"); }
+  virtual uint64_t writev(const Buffer::RawSlice*, uint64_t) { PANIC("not implemented"); }
+
+  virtual void connect(const Network::Address::InstanceConstSharedPtr&) { PANIC("not implemented"); }
+
+  virtual void onAccept(int32_t) { PANIC("not implemented"); };
+  virtual void onClose(int32_t) { PANIC("not implemented"); };
+  virtual void onCancel(int32_t) { PANIC("not implemented"); };
+  virtual void onConnect(int32_t) { PANIC("not implemented"); };
+  virtual void onRead(int32_t) { PANIC("not implemented"); };
+  virtual void onWrite(int32_t, Request*) { PANIC("not implemented"); };
+};
+
+class IoUringWorker : public ThreadLocal::ThreadLocalObject {
+public:
+  virtual ~IoUringWorker() = default;
+
+  virtual void start() PURE;
+
+  virtual void enableSocket(os_fd_t fd) PURE;
+  virtual void disableSocket(os_fd_t fd) PURE;
+  virtual IoUringSocket& addAcceptSocket(os_fd_t fd, IoUringHandler& handler) PURE;
+  virtual IoUringSocket& addServerSocket(os_fd_t fd, IoUringHandler& handler, uint32_t read_buffer_size) PURE;
+  virtual IoUringSocket& addClientSocket(os_fd_t fd, IoUringHandler& handler, uint32_t read_buffer_size) PURE;
+  virtual void closeSocket(os_fd_t fd) PURE;
+  virtual std::unique_ptr<IoUringSocket> removeSocket(os_fd_t) PURE;
+  virtual Event::Dispatcher& dispatcher() PURE;
+  virtual IoUringSocket& getIoUringSocket(os_fd_t fd) PURE;
+
+  virtual Request* submitAcceptRequest(IoUringSocket& socket, sockaddr_storage* remote_addr,
```

Comment:
This is the `user data` pass to the kernel when submit the io uring request, then it will be deallocate here https://github.com/envoyproxy/envoy/pull/24305/files#diff-5a267b3ed3d6b9001c2748d9940424e4582512c97271f77664dd0e126af81160R146, Then I use this to track the request. let me think about if there is better way to manager those pointers in next iteration.

Reply:
Maybe I can add `Request` as a member of IoUringAcceptSocket, since there always only one accept request at one time. Then I needn't allocate/deallocate memory for it.